File: 백준/Silver/1051. 숫자 정사각형/숫자 정사각형.py
import sys

input = sys.stdin.readline

N, M = map(int, input().strip().split())
grid = [input().strip() for _ in range(N)]

side = 1

for i in range(N):
    for j in range(M):
        for size in range(1, min(N - i, M - j)):
            if grid[i][j] == grid[i][j + size] == grid[i + size][j] == grid[i + size][j + size]:
                side = max(side, size + 1)

print(side * side)

File: 백준/Silver/1051. 숫자 정사각형/숫자 정사각형.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        String[] input = br.readLine().trim().split(" ");
        int N = Integer.parseInt(input[0]);
        int M = Integer.parseInt(input[1]);

        String[] grid = new String[N];
        for (int i = 0; i < N; i++) {
            grid[i] = br.readLine().trim();
        }

        int side = 1;

        for(int i = 0; i < N; i++){
            for(int j = 0; j < M; j++){
                for(int size = 1; size + i < N && size + j < M; size++){
                    if (grid[i].charAt(j) == grid[i].charAt(j + size) &&
                        grid[i].charAt(j) == grid[i + size].charAt(j) &&
                        grid[i].charAt(j) == grid[i + size].charAt(j + size)) {
                        side = side > size + 1 ? side : size + 1;
                    }
                }
            }
        }
        System.out.println(side * side);
    }
}